File: app/gululu_immersive.py
"""Parse Gululu author directives into inert EPUB semantic markers."""
from __future__ import annotations

import copy
import html
import re
import urllib.parse
from dataclasses import dataclass


_MANUAL_MUSIC = re.compile(
    r"^\s*<音乐>\s*(.*?)\s*[♪♫]\s*(.*?)\s*</音乐结束>\s*$",
    re.DOTALL,
)
_AUTO_MUSIC = re.compile(
    r"^\s*<自动音乐>\s*(.*?)\s*[♪♫]\s*(.*?)\s*</自动音乐结束>\s*$",
    re.DOTALL,
)
_VFX = re.compile(r"^\s*<特效[:：]\s*(.*?)\s*>\s*$", re.DOTALL)

_VFX_NAMES = {
    "下雨": "rain",
    "雨": "rain",
    "rain": "rain",
    "下雪": "snow",
    "雪": "snow",
    "snow": "snow",
    "打雷": "thunder",
    "雷": "thunder",
    "thunder": "thunder",
    "lightning": "thunder",
    "地震": "quake",
    "震动": "quake",
    "quake": "quake",
    "earthquake": "quake",
    "狂风": "wind",
    "风": "wind",
    "wind": "wind",
    "gale": "wind",
    "停止": "stop",
    "关闭": "stop",
    "stop": "stop",
    "clear": "stop",
}


@dataclass(frozen=True)
class ImmersiveFloor:
    nodes: list[dict]
    vfx: str = ""
    background_update: str | None = None


def safe_https_url(value: object) -> str:
    raw = str(value or "").strip()
    if not raw or len(raw) > 2048 or any(ord(char) < 32 for char in raw):
        return ""
    try:
        parsed = urllib.parse.urlsplit(raw)
    except ValueError:
        return ""
    if (
        parsed.scheme.lower() != "https"
        or not parsed.hostname
        or parsed.username is not None
        or parsed.password is not None
    ):
        return ""
    return raw


def _node_text(node: dict) -> str:
    if str(node.get("type") or "") == "text":
        return str(node.get("text") or "")
    content = node.get("content")
    if not isinstance(content, list):
        return ""
    return "".join(_node_text(child) for child in content if isinstance(child, dict))


def _directive_error(message: str) -> dict:
    return {
        "type": "gululuDirectiveError",
        "attrs": {"message": message},
        "content": [],
    }


def background_attribute(attrs: dict) -> str:
    background = str(attrs.get("gululuBackground") or "")
    if not background:
        return ""
    return f' data-gululu-background-url="{html.escape(background, quote=True)}"'


def render_immersive_node(node_type: str, attrs: dict) -> str | None:
    if node_type == "gululuMusic":
        title = str(attrs.get("title") or "BGM")
        url = html.escape(str(attrs.get("url") or ""), quote=True)
        auto = ' data-gululu-music-auto="true"' if attrs.get("auto") else ""
        label = "自动音乐" if attrs.get("auto") else "音乐"
        return (
            '<p class="gululu-music-row">'
            f'<button type="button" class="gululu-music-cue" '
            f'data-gululu-music-url="{url}"{auto}>'
            f'<span class="gululu-music-kind">{label}</span>'
            f'<span class="gululu-music-title">{html.escape(title)}</span>'
            "</button></p>"
        )
    if node_type == "gululuMusicStop":
        return (
            '<span class="gululu-music-stop" data-gululu-music-stop="true" '
            'role="button" tabindex="0" aria-label="停止音乐">&#9632;</span>'
        )
    if node_type == "gululuBackgroundClear":
        return (
            '<span class="gululu-immersive-marker" '
            'data-gululu-background-clear="true" aria-hidden="true"></span>'
        )
    if node_type == "gululuDirectiveError":
        message = html.escape(str(attrs.get("message") or "沉浸指令不可用"))
        return f'<p class="gululu-directive-error">[{message}]</p>'
    return None


def prepare_immersive_floor(nodes: object) -> ImmersiveFloor:
    """Recognize complete directive paragraphs without mutating the API snapshot."""
    if not isinstance(nodes, list):
        return ImmersiveFloor([])

    output: list[dict] = []
    vfx = ""
    in_background = False
    background_update: str | None = None

    for source in nodes:
        if not isinstance(source, dict):
            continue
        node = copy.deepcopy(source)
        node_type = str(node.get("type") or "")
        text = _node_text(node).strip() if node_type == "paragraph" else ""

        music = _AUTO_MUSIC.fullmatch(text) or _MANUAL_MUSIC.fullmatch(text)
        if music:
            title = music.group(1).strip() or "BGM"
            url = safe_https_url(music.group(2))
            if not url:
                output.append(_directive_error(f"音乐链接不可用：{title}"))
            else:
                output.append({
                    "type": "gululuMusic",
                    "attrs": {
                        "title": title,
                        "url": url,
                        "auto": bool(_AUTO_MUSIC.fullmatch(text)),
                    },
                    "content": [],
                })
            continue

        if text == "<停止音乐>":
            output.append({"type": "gululuMusicStop", "attrs": {}, "content": []})
            continue

        effect_match = _VFX.fullmatch(text)
        if effect_match:
            requested = effect_match.group(1).strip()
            mapped = _VFX_NAMES.get(requested.lower(), "")
            if mapped:
                if not vfx:
                    vfx = mapped
            else:
                output.append(_directive_error(f"暂不支持的特效：{requested or '空'}"))
            continue

        if text == "<背景>":
            in_background = True
            continue
        if text == "</背景>":
            if in_background:
                in_background = False
                continue
        if text in {"<移除背景>", "<清除背景>", "<恢复背景>"}:
            output.append({"type": "gululuBackgroundClear", "attrs": {}, "content": []})
            background_update = ""
            continue

        if in_background and node_type == "image":
            attrs = node.get("attrs") if isinstance(node.get("attrs"), dict) else {}
            background_url = safe_https_url(attrs.get("src"))
            if background_url:
                attrs["gululuBackground"] = background_url
                node["attrs"] = attrs
                background_update = background_url
            else:
                output.append(_directive_error("背景图片链接不可用"))
        output.append(node)

    if in_background:
        output.append(_directive_error("背景指令缺少结束标记"))
    return ImmersiveFloor(output, vfx=vfx, background_update=background_update)

File: android/app/src/main/java/io/github/gighi947/ankeshelf/data/GululuImmersive.kt
package io.github.gighi947.ankeshelf.data

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/** 一层的沉浸解析结果（nodes 已含助手协议转换；vfx/背景交给宿主层呈现）。 */
data class ImmersiveFloor(
    val nodes: List<JsonObject>,
    val vfx: String = "",
    val backgroundUpdate: String? = null,
)

/**
 * 骨碌碌作者沉浸指令 → 惰性 EPUB 语义标记（Kotlin 版 `app/gululu_immersive.py`）。
 *
 * 红线（与桌面一致）：只接受**无凭据 HTTPS** 外链；指令段落整体识别后替换为
 * `data-*` 标记节点，不在运行时改写正文；视效每层只取第一个有效值。
 */
object GululuImmersive {

    private val MANUAL_MUSIC = Regex(
        "^\\s*<音乐>\\s*(.*?)\\s*[♪♫]\\s*(.*?)\\s*</音乐结束>\\s*$",
        RegexOption.DOT_MATCHES_ALL,
    )
    private val AUTO_MUSIC = Regex(
        "^\\s*<自动音乐>\\s*(.*?)\\s*[♪♫]\\s*(.*?)\\s*</自动音乐结束>\\s*$",
        RegexOption.DOT_MATCHES_ALL,
    )
    private val VFX = Regex("^\\s*<特效[:：]\\s*(.*?)\\s*>\\s*$", RegexOption.DOT_MATCHES_ALL)

    private val VFX_NAMES = mapOf(
        "下雨" to "rain", "雨" to "rain", "rain" to "rain",
        "下雪" to "snow", "雪" to "snow", "snow" to "snow",
        "打雷" to "thunder", "雷" to "thunder", "thunder" to "thunder", "lightning" to "thunder",
        "地震" to "quake", "震动" to "quake", "quake" to "quake", "earthquake" to "quake",
        "狂风" to "wind", "风" to "wind", "wind" to "wind", "gale" to "wind",
        "停止" to "stop", "关闭" to "stop", "stop" to "stop", "clear" to "stop",
    )

    private const val DIRECTIVE_PADDING = " \t\r\n\u00a0\u200b\ufeff\u3000"
    private val BACKGROUND_CLEAR = setOf("<移除背景>", "<清除背景>", "<恢复背景>")

    /** 只接受无凭据 HTTPS（无用户名/密码、无控制字符、长度 ≤2048）；否则返回空串。 */
    fun safeHttpsUrl(value: String?): String {
        val raw = value?.trim().orEmpty()
        if (raw.isEmpty() || raw.length > 2048 || raw.any { it.code < 32 }) return ""
        val uri = runCatching { java.net.URI(raw) }.getOrNull() ?: return ""
        if (uri.scheme?.lowercase() != "https") return ""
        if (uri.host.isNullOrEmpty()) return ""
        if (uri.userInfo != null) return ""
        return raw
    }

    /** 图片节点的背景标记属性（由 [prepareImmersiveFloor] 写入 attrs 后渲染时读出）。 */
    fun backgroundAttribute(attrs: JsonObject): String {
        val background = (attrs["gululuBackground"] as? JsonPrimitive)?.content.orEmpty()
        if (background.isEmpty() || background == "null") return ""
        return " data-gululu-background-url=\"${GululuAst.escape(background)}\""
    }

    /**
     * 识别完整的指令段落（不修改接口快照）：音乐/自动音乐/停止音乐、视效、
     * 背景区间与清除背景；其余节点原样保留，最后统一过一遍助手协议。
     */
    fun prepareImmersiveFloor(nodes: JsonElement?): ImmersiveFloor {
        val list = nodes as? JsonArray ?: return ImmersiveFloor(emptyList())
        val output = mutableListOf<JsonElement>()
        var vfx = ""
        var inBackground = false
        var backgroundUpdate: String? = null

        for (source in list.filterIsInstance<JsonObject>()) {
            val nodeType = (source["type"] as? JsonPrimitive)?.content.orEmpty()
            val text = if (nodeType == "paragraph") {
                GululuAssistant.nodeText(source).trim { it in DIRECTIVE_PADDING }
            } else {
                ""
            }

            val auto = AUTO_MUSIC.matchEntire(text)
            val music = auto ?: MANUAL_MUSIC.matchEntire(text)
            if (music != null) {
                val title = music.groupValues[1].trim().ifEmpty { "BGM" }
                val url = safeHttpsUrl(music.groupValues[2])
                if (url.isEmpty()) {
                    output.add(GululuAssistant.directiveError("音乐链接不可用：$title"))
                } else {
                    output.add(
                        JsonObject(
                            mapOf(
                                "type" to JsonPrimitive("gululuMusic"),
                                "attrs" to buildJsonObject {
                                    put("title", title)
                                    put("url", url)
                                    put("auto", auto != null)
                                },
                                "content" to JsonArray(emptyList()),
                            ),
                        ),
                    )
                }
                continue
            }

            if (text == "<停止音乐>") {
                output.add(marker("gululuMusicStop"))
                continue
            }

            val effect = VFX.matchEntire(text)
            if (effect != null) {
                val requested = effect.groupValues[1].trim()
                val mapped = VFX_NAMES[requested.lowercase()].orEmpty()
                if (mapped.isNotEmpty()) {
                    if (vfx.isEmpty()) vfx = mapped
                } else {
                    output.add(
                        GululuAssistant.directiveError(
                            "暂不支持的特效：${requested.ifEmpty { "空" }}",
                        ),
                    )
                }
                continue
            }

            if (text == "<背景>") {
                inBackground = true
                continue
            }
            if (text == "</背景>" && inBackground) {
                inBackground = false
                continue
            }
            if (text in BACKGROUND_CLEAR) {
                output.add(marker("gululuBackgroundClear"))
                backgroundUpdate = ""
                continue
            }

            var node = source
            if (inBackground && nodeType == "image") {
                val attrs = source["attrs"] as? JsonObject ?: JsonObject(emptyMap())
                val backgroundUrl = safeHttpsUrl((attrs["src"] as? JsonPrimitive)?.content)
                if (backgroundUrl.isNotEmpty()) {
                    val patched = JsonObject(
                        attrs.toMutableMap().apply {
                            put("gululuBackground", JsonPrimitive(backgroundUrl))
                        },
                    )
                    node = JsonObject(source.toMutableMap().apply { put("attrs", patched) })
                    backgroundUpdate = backgroundUrl
                } else {
                    output.add(GululuAssistant.directiveError("背景图片链接不可用"))
                }
            }
            output.add(node)
        }

        if (inBackground) output.add(GululuAssistant.directiveError("背景指令缺少结束标记"))
        return ImmersiveFloor(
            nodes = GululuAssistant.prepareAssistantNodes(output),
            vfx = vfx,
            backgroundUpdate = backgroundUpdate,
        )
    }

    /** 渲染扩展（对齐 `render_immersive_node`）。 */
    fun renderer(): GululuNodeRenderer = GululuNodeRenderer { nodeType, attrs, _, _, _ ->
        val escape = GululuAst::escape
        when (nodeType) {
            "gululuMusic" -> {
                val title = (attrs["title"] as? JsonPrimitive)?.content?.ifEmpty { "BGM" } ?: "BGM"
                val url = escape((attrs["url"] as? JsonPrimitive)?.content.orEmpty())
                val isAuto = (attrs["auto"] as? JsonPrimitive)?.let {
                    it.content == "true" || it.content == "1"
                } ?: false
                val auto = if (isAuto) " data-gululu-music-auto=\"true\"" else ""
                val label = if (isAuto) "自动音乐" else "音乐"
                "<p class=\"gululu-music-row\">" +
                    "<button type=\"button\" class=\"gululu-music-cue\" " +
                    "data-gululu-music-url=\"$url\"$auto>" +
                    "<span class=\"gululu-music-kind\">$label</span>" +
                    "<span class=\"gululu-music-title\">${escape(title)}</span>" +
                    "</button></p>"
            }
            "gululuMusicStop" ->
                "<span class=\"gululu-music-stop\" data-gululu-music-stop=\"true\" " +
                    "role=\"button\" tabindex=\"0\" aria-label=\"停止音乐\">&#9632;</span>"
            "gululuBackgroundClear" ->
                "<span class=\"gululu-immersive-marker\" " +
                    "data-gululu-background-clear=\"true\" aria-hidden=\"true\"><wbr/></span>"
            "gululuDirectiveError" -> {
                val message = escape(
                    (attrs["message"] as? JsonPrimitive)?.content?.ifEmpty { "沉浸指令不可用" }
                        ?: "沉浸指令不可用",
                )
                "<p class=\"gululu-directive-error\">[$message]</p>"
            }
            else -> null
        }
    }

    private fun marker(type: String): JsonObject = JsonObject(
        mapOf(
            "type" to JsonPrimitive(type),
            "attrs" to JsonObject(emptyMap()),
            "content" to JsonArray(emptyList()),
        ),
    )
}
